// pixi.js/lib/scene/text-html/utils/textStyleToCSS.js
'use strict';

var Color = require('../../../color/Color.js');
var TextStyle = require('../../text/TextStyle.js');

"use strict";
function textStyleToCSS(style) {
  const stroke = style._stroke;
  const fill = style._fill;
  const color = Color.Color.shared.setValue(fill.color).setAlpha(fill.alpha ?? 1).toHexa();
  const cssStyleString = [
    `color: ${color}`,
    `font-size: ${style.fontSize}px`,
    `font-family: ${style.fontFamily}`,
    `font-weight: ${style.fontWeight}`,
    `font-style: ${style.fontStyle}`,
    `font-variant: ${style.fontVariant}`,
    `letter-spacing: ${style.letterSpacing}px`,
    `text-align: ${style.align}`,
    `padding: ${style.padding}px`,
    `white-space: ${style.whiteSpace === "pre" && style.wordWrap ? "pre-wrap" : style.whiteSpace}`,
    ...style.lineHeight ? [`line-height: ${style.lineHeight}px`] : [],
    ...style.wordWrap ? [
      `word-break: ${style.breakWords ? "break-word" : "normal"}`,
      `max-width: ${style.wordWrapWidth}px`
    ] : [],
    ...stroke ? [strokeToCSS(stroke)] : [],
    ...style.dropShadow ? [dropShadowToCSS(style.dropShadow)] : [],
    ...style.cssOverrides
  ].join(";");
  const cssStyles = [`div { ${cssStyleString} }`];
  tagStyleToCSS(style.tagStyles, cssStyles);
  return cssStyles.join(" ");
}
function dropShadowToCSS(dropShadowStyle) {
  const dropshadowStyle = { ...dropShadowStyle };
  const color = Color.Color.shared.setValue(dropshadowStyle.color).setAlpha(dropshadowStyle.alpha ?? 1).toHexa();
  const x = Math.round(Math.cos(dropshadowStyle.angle) * dropshadowStyle.distance);
  const y = Math.round(Math.sin(dropshadowStyle.angle) * dropshadowStyle.distance);
  const position = `${x}px ${y}px`;
  if (dropshadowStyle.blur > 0) {
    return `text-shadow: ${position} ${dropshadowStyle.blur}px ${color}`;
  }
  return `text-shadow: ${position} ${color}`;
}
function strokeToCSS(stroke) {
  const color = Color.Color.shared.setValue(stroke.color).setAlpha(stroke.alpha ?? 1).toHexa();
  return [
    `-webkit-text-stroke-width: ${stroke.width}px`,
    `-webkit-text-stroke-color: ${color}`,
    `text-stroke-width: ${stroke.width}px`,
    `text-stroke-color: ${color}`,
    "paint-order: stroke"
  ].join(";");
}
const templates = {
  fontSize: `font-size: {{VALUE}}px`,
  fontFamily: `font-family: {{VALUE}}`,
  fontWeight: `font-weight: {{VALUE}}`,
  fontStyle: `font-style: {{VALUE}}`,
  fontVariant: `font-variant: {{VALUE}}`,
  letterSpacing: `letter-spacing: {{VALUE}}px`,
  align: `text-align: {{VALUE}}`,
  padding: `padding: {{VALUE}}px`,
  whiteSpace: `white-space: {{VALUE}}`,
  lineHeight: `line-height: {{VALUE}}px`,
  wordWrapWidth: `max-width: {{VALUE}}px`
};
const transform = {
  fill: (value) => `color: ${Color.Color.shared.setValue(value).toHexa()}`,
  breakWords: (value) => `word-break: ${value ? "break-all" : "normal"}`,
  stroke: strokeToCSS,
  dropShadow: (value) => {
    if (value === true) {
      return dropShadowToCSS(TextStyle.TextStyle.defaultDropShadow);
    }
    if (value && typeof value === "object") {
      return dropShadowToCSS({ ...TextStyle.TextStyle.defaultDropShadow, ...value });
    }
    return "";
  }
};
function tagStyleToCSS(tagStyles, out) {
  for (const i in tagStyles) {
    const tagStyle = tagStyles[i];
    const cssTagStyle = [];
    for (const j in tagStyle) {
      if (transform[j]) {
        cssTagStyle.push(transform[j](tagStyle[j]));
      } else if (templates[j]) {
        cssTagStyle.push(templates[j].replace("{{VALUE}}", tagStyle[j]));
      }
    }
    out.push(`${i} { ${cssTagStyle.join(";")} }`);
  }
}

exports.textStyleToCSS = textStyleToCSS;
